### ISSUE TITLE
RFC: Improvements to the script & go server

### ISSUE THREAD

**Issue by intentionally-left-nil** [samm-git/aws-vpn-client#17]
# TL;DR
I made some improvements that I'd be happy to backport to this repository if there's interest. Please let me know which pieces are valued and I'll get to work.

Take a look at
https://github.com/AnilRedshift/awsvpn-saml/blob/main/main.go
and https://github.com/AnilRedshift/awsvpn-saml/blob/main/awsvpn
Usage of my codebase: Install (on arch) with makepkg -i . Then, just run `awsvpn configure` to set your variables, and then run `awsvpn` to run the vpn

## Prefer the existing environment variables
Basically anyone who uses the current aws-connect.sh has to modify the first few lines to set the environment variables. You could make this useable without any code changes with something like this:
```
VPN_CONF="${VPN_CONF:=$HOME/aws.ovpn}"
VPN_PORT="${VPN_PORT:=443}"
VPN_PROTO="${VPN_PROTO:=udp}"
```

which only overrides the values if they're not already set on the command line.

## Use a configuration file
My [script](https://github.com/AnilRedshift/awsvpn-saml/blob/main/awsvpn) models after the aws-cli binary and has a `awsvpn configure` option. It creates a file (awsvpn.env) which it then loads.
```
if [ -f "$config_file" ]; then
  echo "sourcing $config_file"
  export $(egrep -v '^#' "$config_file" | xargs)
else
  echo "No env file found"
fi
```

## Start/stop the go server from within the script
Instead of requiring the user to manually start/stop the go server, I'm doing it automatically:
```
coproc "$bin_dir/awsvpnserver"
srv_pid=$!
exec 3>&"${COPROC[0]}"
clean_up () {
  arg=$?
  echo "Stopping the server for SAML authentication ($srv_pid)"
  kill -9 "$srv_pid" > /dev/null 2>&1 || true
  exit "$arg"
}
```

## Use stdout instead of SAMLresponse.txt to capture the saml token.
I modified the go server to write to stdout `fmt.Println(saml)` instead of writing to a file. Then, using `coproc`, I just read that directly from the script, instead of having this file lying around

## Stop the go server once authenticated
Pretty simple idea - this doesn't need to be a long-lived process, so I nuke it once the connection succeeds:
```
http.HandleFunc("/", func(rw http.ResponseWriter, r *http.Request) {
		if saml := r.FormValue("SAMLResponse"); saml != "" {
			rw.WriteHeader(http.StatusOK)
			rw.Write([]byte("succeeded"))
			go func() {
				time.Sleep(time.Second * 5) // Let the POST request return to the caller successfully before returning the SAML token
				fmt.Println(saml)
				ctx, onComplete := context.WithTimeout(context.Background(), time.Second*5)
				srv.Shutdown(ctx)
				onComplete()
			}()
		} else {
			rw.WriteHeader(http.StatusBadRequest)
		}
	})
```

## Fix quoting in the bash script
This line is bugged:
```
sudo bash -c "$OVPN_BIN --config "${OVPN_CONF}" \
    --verb 3 --auth-nocache --inactive 3600 \
    --proto "$PROTO" --remote $SRV $PORT \
    --script-security 2 \
    --route-up '/usr/bin/env rm saml-response.txt' \
    --auth-user-pass <( printf \"%s\n%s\n\" \"N/A\" \"CRV1::${VPN_SID}::$(cat saml-response.txt)\" )"
```

Because you have sudo bash -c "" (with the quotes around the bash command), you need to escape every quote within the script. OVPN_CONF and PROTO are not escaped, causing them to be literals.

## Remove broken openvpn tests
The aws patch breaks the openvpn test suite (because it's a dumb hack).
This patch removes the broken tests, so you can run `make check` on openvpn and verify the other behavior works:
https://github.com/AnilRedshift/awsvpn-saml/blob/main/skip-broken-tests-2.5.1.patch

## Passthrough remaining arguments to openvpn
Added `"$@"' to the openvpn binary to pass through any custom arguments via the script

**Comment by samm-git** [comment]
If issue is still relevant, please open it on https://github.com/aws-vpn-client/aws-vpn-client
